### PR TITLE
Addition of SDFGI for open world global illumination

### PR DESCRIPTION
**Pull request by reduz**
image
image

-Added open world global illumination based on signed distance field cascades, sphere tracing and voxel occlusion.
-Also Moved GI to a deferred pass
-Removed reliance on large arrays for GIProbe

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,2 @@
+all:
```
Comment (fire) [inline Makefile:1]: Make is not necessary.

File: servers/rendering/rasterizer_rd/rasterizer_scene_rd.h [hunk 2]
```diff
@@ -764,10 +787,389 @@ class RasterizerSceneRD : public RasterizerScene {
...
+		RID debug_probes_shader_version;
+
+		RenderPipelineVertexFormatCacheRD debug_probes_pipeline[PROBE_DEBUG_MAX];
+
+		struct Light {
+			float color[3];
+			float energy;
+
+			float direction[3];
+			uint has_shadow;
```
Comment (fire) [inline servers/rendering/rasterizer_rd/rasterizer_scene_rd.h:1004]: has_shadow.. Is this a boolean?

error C3646: 'has_shadow': unknown override specifier

File: servers/rendering/rasterizer_rd/rasterizer_scene_rd.h [hunk 3]
```diff
@@ -764,10 +787,389 @@ class RasterizerSceneRD : public RasterizerScene {
...
+			int32_t probe_axis_size;
+		};
+
+		SdfgiDebugProbesShaderRD debug_probes;
+		RID debug_probes_shader;
+		RID debug_probes_shader_version;
+
+		RenderPipelineVertexFormatCacheRD debug_probes_pipeline[PROBE_DEBUG_MAX];
+
+		struct Light {
```
Comment (fire) [inline servers/rendering/rasterizer_rd/rasterizer_scene_rd.h:999]: I believe the new core style is init on the constructor so like float radius = 0;

**Review by akien-mga — APPROVED** [review]
A thorough review will take a while so I suggest that we merge as is (once CI passes) and take the time needed to do a post-merge review to see what needs to be changed.

File: core/math/basis.cpp [hunk 4]
```diff
@@ -773,7 +773,7 @@ Basis::operator String() const {
 				mtx += ", ";
 			}
 
-			mtx += rtos(elements[i][j]);
+			mtx += rtos(elements[j][i]); //matrix is stored transposed for performance, so print it transposed
```
Comment (aaronfranke) [inline core/math/basis.cpp:776]: @reduz This change doesn't seem to be related to SDFGI, was there a reason why it was changed in this PR?

Regardless, #34668 makes Basis and Transform(2D) print 2-dimensionally, which avoids confusion.

File: servers/rendering/rasterizer_rd/shaders/scene_high_end.glsl [hunk 5]
```diff
@@ -2376,6 +2497,93 @@ FRAGMENT_SHADER_CODE
...
+
+		vec3 cam_normal = mat3(scene_data.camera_matrix) * normal;
+
+		for (uint i = 0; i < 6; i++) {
+			if (dot(cam_normal, aniso_dir[i]) > 0.001) {
+				facing_bits |= (1 << i);
+			}
+		}
+
+		imageAtomicOr(geom_facing_grid, grid_pos, facing_bits); //store facing bits
```
Comment (bruvzg) [inline servers/rendering/rasterizer_rd/shaders/scene_high_end.glsl:2530]: This line fails on macOS with following error:

```
[mvk-error] VK_ERROR_INITIALIZATION_FAILED: Shader library compile failed (Error code 3):
Compilation failed:

program_source:390:87: error: array subscript is not an integer
    uint _331 = atomic_fetch_or_explicit((device atomic_uint*)&geom_facing_grid_atomic[grid_pos], facing_bits, memory_order_relaxed);
                                                                                      ^~~~~~~~~
```

Apparently, Metal doesn't support atomic ops on texture storage in fragment shaders at all (https://github.com/KhronosGroup/SPIRV-Cross/issues/835).

File: servers/rendering/rasterizer_rd/rasterizer_scene_rd.cpp [hunk 6]
```diff
@@ -162,37 +162,1572 @@ void RasterizerSceneRD::_create_reflection_fast_filter(ReflectionData &rd, bool
...
+			tf_render.depth /= 2;
+
+			sdfgi->render_sdf_half[0] = RD::get_singleton()->texture_create(tf_render, RD::TextureView());
+			sdfgi->render_sdf_half[1] = RD::get_singleton()->texture_create(tf_render, RD::TextureView());
+		}
+
+		RD::TextureFormat tf_occlusion = tf_sdf;
+		tf_occlusion.format = RD::DATA_FORMAT_R16_UINT;
+		tf_occlusion.shareable_formats.push_back(RD::DATA_FORMAT_R16_UINT);
+		tf_occlusion.shareable_formats.push_back(RD::DATA_FORMAT_R4G4B4A4_UNORM_PACK16);
```
Comment (bruvzg) [inline servers/rendering/rasterizer_rd/rasterizer_scene_rd.cpp:322]: Fails on macOS:
```
[mvk-error] VK_ERROR_FORMAT_NOT_SUPPORTED: VkFormat VK_FORMAT_R4G4B4A4_UNORM_PACK16 is not supported on this device.
```

Comment (bruvzg) [reply servers/rendering/rasterizer_rd/rasterizer_scene_rd.cpp:322]: Supported formats for the reference:

<details>
  <summary>Click to expand</summary>

```
	Formats: count = 50
		SurfaceFormat[0]:
			format = FORMAT_B8G8R8A8_UNORM
			colorSpace = COLOR_SPACE_SRGB_NONLINEAR_KHR
		SurfaceFormat[1]:
			format = FORMAT_B8G8R8A8_SRGB
			colorSpace = COLOR_SPACE_SRGB_NONLINEAR_KHR
		SurfaceFormat[2]:
			format = FORMAT_R16G16B16A16_SFLOAT
			colorSpace = COLOR_SPACE_SRGB_NONLINEAR_KHR
		SurfaceFormat[3]:
			format = FORMAT_A2B10G10R10_UNORM_PACK32
			colorSpace = COLOR_SPACE_SRGB_NONLINEAR_KHR
		SurfaceFormat[4]:
			format = FORMAT_A2R10G10B10_UNORM_PACK32
			colorSpace = COLOR_SPACE_SRGB_NONLINEAR_KHR
		SurfaceFormat[5]:
			format = FORMAT_B8G8R8A8_UNORM
			colorSpace = COLOR_SPACE_DISPLAY_P3_NONLINEAR_EXT
		SurfaceFormat[6]:
			format = FORMAT_B8G8R8A8_SRGB
			colorSpace = COLOR_SPACE_DISPLAY_P3_NONLINEAR_EXT
		SurfaceFormat[7]:
			format = FORMAT_R16G16B16A16_SFLOAT
			colorSpace = COLOR_SPACE_DISPLAY_P3_NONLINEAR_EXT
		SurfaceFormat[8]:
			format = FORMAT_A2B10G10R10_UNORM_PACK32
			colorSpace = COLOR_SPACE_DISPLAY_P3_NONLINEAR_EXT
		SurfaceFormat[9]:
			format = FORMAT_A2R10G10B10_UNORM_PACK32
			colorSpace = COLOR_SPACE_DISPLAY_P3_NONLINEAR_EXT
		SurfaceFormat[10]:
			format = FORMAT_B8G8R8A8_UNORM
			colorSpace = COLOR_SPACE_DCI_P3_NONLINEAR_EXT
		SurfaceFormat[11]:
			format = FORMAT_B8G8R8A8_SRGB
			colorSpace = COLOR_SPACE_DCI_P3_NONLINEAR_EXT
		SurfaceFormat[12]:
			format = FORMAT_R16G16B16A16_SFLOAT
			colorSpace = COLOR_SPACE_DCI_P3_NONLINEAR_EXT
		SurfaceFormat[13]:
			format = FORMAT_A2B10G10R10_UNORM_PACK32
			colorSpace = COLOR_SPACE_DCI_P3_NONLINEAR_EXT
		SurfaceFormat[14]:
			format = FORMAT_A2R10G10B10_UNORM_PACK32
			colorSpace = COLOR_SPACE_DCI_P3_NONLINEAR_EXT
		SurfaceFormat[15]:
			format = FORMAT_B8G8R8A8_UNORM
			colorSpace = COLOR_SPACE_BT709_NONLINEAR_EXT
		SurfaceFormat[16]:
			format = FORMAT_B8G8R8A8_SRGB
			colorSpace = COLOR_SPACE_BT709_NONLINEAR_EXT
		SurfaceFormat[17]:
			format = FORMAT_R16G16B16A16_SFLOAT
			colorSpace = COLOR_SPACE_BT709_NONLINEAR_EXT
		SurfaceFormat[18]:
			format = FORMAT_A2B10G10R10_UNORM_PACK32
			colorSpace = COLOR_SPACE_BT709_NONLINEAR_EXT
		SurfaceFormat[19]:
			format = FORMAT_A2R10G10B10_UNORM_PACK32
			colorSpace = COLOR_SPACE_BT709_NONLINEAR_EXT
		SurfaceFormat[20]:
			format = FORMAT_B8G8R8A8_UNORM
			colorSpace = COLOR_SPACE_ADOBERGB_NONLINEAR_EXT
		SurfaceFormat[21]:
			format = FORMAT_B8G8R8A8_SRGB
			colorSpace = COLOR_SPACE_ADOBERGB_NONLINEAR_EXT
		SurfaceFormat[22]:
			format = FORMAT_R16G16B16A16_SFLOAT
			colorSpace = COLOR_SPACE_ADOBERGB_NONLINEAR_EXT
		SurfaceFormat[23]:
			format = FORMAT_A2B10G10R10_UNORM_PACK32
			colorSpace = COLOR_SPACE_ADOBERGB_NONLINEAR_EXT
		SurfaceFormat[24]:
			format = FORMAT_A2R10G10B10_UNORM_PACK32
			colorSpace = COLOR_SPACE_ADOBERGB_NONLINEAR_EXT
		SurfaceFormat[25]:
			format = FORMAT_B8G8R8A8_UNORM
			colorSpace = COLOR_SPACE_PASS_THROUGH_EXT
		SurfaceFormat[26]:
			format = FORMAT_B8G8R8A8_SRGB
			colorSpace = COLOR_SPACE_PASS_THROUGH_EXT
		SurfaceFormat[27]:
			format = FORMAT_R16G16B16A16_SFLOAT
			colorSpace = COLOR_SPACE_PASS_THROUGH_EXT
		SurfaceFormat[28]:
			format = FORMAT_A2B10G10R10_UNORM_PACK32
			colorSpace = COLOR_SPACE_PASS_THROUGH_EXT
		SurfaceFormat[29]:
			format = FORMAT_A2R10G10B10_UNORM_PACK32
			colorSpace = COLOR_SPACE_PASS_THROUGH_EXT
		SurfaceFormat[30]:
			format = FORMAT_B8G8R8A8_UNORM
			colorSpace = COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT
		SurfaceFormat[31]:
			format = FORMAT_B8G8R8A8_SRGB
			colorSpace = COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT
		SurfaceFormat[32]:
			format = FORMAT_R16G16B16A16_SFLOAT
			colorSpace = COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT
		SurfaceFormat[33]:
			format = FORMAT_A2B10G10R10_UNORM_PACK32
			colorSpace = COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT
		SurfaceFormat[34]:
			format = FORMAT_A2R10G10B10_UNORM_PACK32
			colorSpace = COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT
		SurfaceFormat[35]:
			format = FORMAT_B8G8R8A8_UNORM
			colorSpace = COLOR_SPACE_EXTENDED_SRGB_NONLINEAR_EXT
		SurfaceFormat[36]:
			format = FORMAT_B8G8R8A8_SRGB
			colorSpace = COLOR_SPACE_EXTENDED_SRGB_NONLINEAR_EXT
		SurfaceFormat[37]:
			format = FORMAT_R16G16B16A16_SFLOAT
			colorSpace = COLOR_SPACE_EXTENDED_SRGB_NONLINEAR_EXT
		SurfaceFormat[38]:
			format = FORMAT_A2B10G10R10_UNORM_PACK32
			colorSpace = COLOR_SPACE_EXTENDED_SRGB_NONLINEAR_EXT
		SurfaceFormat[39]:
			format = FORMAT_A2R10G10B10_UNORM_PACK32
			colorSpace = COLOR_SPACE_EXTENDED_SRGB_NONLINEAR_EXT
		SurfaceFormat[40]:
			format = FORMAT_B8G8R8A8_UNORM
			colorSpace = COLOR_SPACE_DISPLAY_P3_LINEAR_EXT
		SurfaceFormat[41]:
			format = FORMAT_B8G8R8A8_SRGB
			colorSpace = COLOR_SPACE_DISPLAY_P3_LINEAR_EXT
		SurfaceFormat[42]:
			format = FORMAT_R16G16B16A16_SFLOAT
			colorSpace = COLOR_SPACE_DISPLAY_P3_LINEAR_EXT
		SurfaceFormat[43]:
			format = FORMAT_A2B10G10R10_UNORM_PACK32
			colorSpace = COLOR_SPACE_DISPLAY_P3_LINEAR_EXT
		SurfaceFormat[44]:
			format = FORMAT_A2R10G10B10_UNORM_PACK32
			colorSpace = COLOR_SPACE_DISPLAY_P3_LINEAR_EXT
		SurfaceFormat[45]:
			format = FORMAT_B8G8R8A8_UNORM
			colorSpace = COLOR_SPACE_BT2020_LINEAR_EXT
		SurfaceFormat[46]:
			format = FORMAT_B8G8R8A8_SRGB
			colorSpace = COLOR_SPACE_BT2020_LINEAR_EXT
		SurfaceFormat[47]:
			format = FORMAT_R16G16B16A16_SFLOAT
			colorSpace = COLOR_SPACE_BT2020_LINEAR_EXT
		SurfaceFormat[48]:
			format = FORMAT_A2B10G10R10_UNORM_PACK32
			colorSpace = COLOR_SPACE_BT2020_LINEAR_EXT
		SurfaceFormat[49]:
			format = FORMAT_A2R10G10B10_UNORM_PACK32
			colorSpace = COLOR_SPACE_BT2020_LINEAR_EXT
```

</details>
